### ISSUE TITLE
wrong ansible inventory when using ipv6. "the field port has an invalid value"

### ISSUE THREAD

**Issue by pescobar** [gc3-uzh-ch/elasticluster#478]
when testing elasticluster in a openstack cloud using ipv6 I got this error for some nodes during the `gather facts` task in ansible:

```
fatal: [compute017]: FAILED! => {"failed": true, "msg": "the field 'port' has an invalid value (620:5ca1:2f0:f816:3eff:fedb:34ae), and could not be converted to an int. Error was: invalid literal for int() with base 10: '620:5ca1:2f0:f816:3eff:fedb:34ae'"}
```

with the help of @riccardomurri we noticed that the ansible inventory is not properly created. This is a example of a wrong inventory:

```
[slurm_master]
frontend001 ansible_host=10.0.0.46 ansible_user=centos

[slurm_worker]
compute001 ansible_host=10.0.0.66 ansible_user=centos
compute002 ansible_host=10.0.0.55 ansible_user=centos
compute003 ansible_host=10.0.0.70 ansible_user=centos
compute004 ansible_host=10.0.0.56 ansible_user=centos
compute005 ansible_host=10.0.0.68 ansible_user=centos
compute006 ansible_host=10.0.0.3 ansible_user=centos
compute007 ansible_host=10.0.0.75 ansible_user=centos
compute008 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fe68:17e2
compute009 ansible_host=10.0.0.67 ansible_user=centos
compute010 ansible_host=10.0.0.58 ansible_user=centos
compute011 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fe97:165e
compute012 ansible_host=10.0.0.35 ansible_user=centos
compute013 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fe4a:141c
compute014 ansible_host=10.0.0.82 ansible_user=centos
compute015 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fe17:1dd9
compute016 ansible_host=10.0.0.79 ansible_user=centos
compute017 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fedb:34ae
compute018 ansible_host=10.0.0.63 ansible_user=centos
compute019 ansible_host=2001 ansible_user=centos ansible_port=620:5ca1:2f0:f816:3eff:fed8:f91f
compute020 ansible_host=10.0.0.78 ansible_user=centos
compute021 ansible_host=10.0.0.84 ansible_user=centos
compute022 ansible_host=10.0.0.80 ansible_user=centos
compute023 ansible_host=10.0.0.51 ansible_user=centos
compute024 ansible_host=10.0.0.72 ansible_user=centos
compute025 ansible_host=10.0.0.96 ansible_user=centos
compute026 ansible_host=10.0.0.91 ansible_user=centos
```

**Comment by pescobar** [comment]
after testing the fix in https://github.com/riccardomurri/elasticluster/tree/fix/inventory-with-ipv6 I get this error

```
2017-08-16 18:13:07 elasticluster-admin.novalocal gc3.elasticluster[21486] ERROR The cluster hosts are up and running, but Ansible failed to set the cluster up: global name 'IPV6_RE' is not defined
```

And this (empty) inventory file

```[slurm_master]
frontend001 ansible_host=10.0.0.56 ansible_user=centos

[slurm_worker]
```

**Comment by riccardomurri** [comment]
It's actually not needed to fully configure the cluster to test if the bug is there.
You can just run this snippet of shell code over and over::

    elasticluster start --no-setup mycluster
    if egrep $HOME/.elasticluster/storage/mycluster.inventory 'ansible_port=[[:xdigit:]]+:[[:xdigit:]]+'; then
      echo "BUG!" 2>&1
    fi
    elasticluster stop -y mycluster

**Comment by pescobar** [comment]
after doing a `git pull origin fix/inventory-with-ipv6` you snippet was executed correctly without errors. 

In the inventory file I could only find ipv4 address

**Comment by riccardomurri** [comment]
Could you run the snippet a few times over?  Since the failure is not deterministically reproducible we need to rely on statistics ;-)

**Comment by pescobar** [comment]
@riccardomurri I have modified the snippet like this so it works correctly (changed the arguments order for grep)

```
cluster_name="slurm_centos7"

elasticluster start --no-setup ${cluster_name}
if egrep 'ansible_port=[[:xdigit:]]+:[[:xdigit:]]+' $HOME/.elasticluster/storage/${cluster_name}.inventory ; then
  echo "BUG!" 2>&1
fi
elasticluster stop -y ${cluster_name}
```
I don't get any error but I have noticed that even if I am booting a cluster with 26 nodes I only get three entries in the inventory file . e.g.

```
[centos@elasticluster-admin ~]$ cat ~/.elasticluster/storage/slurm_centos7.inventory 

[slurm_master]
frontend001 ansible_host=10.0.0.63 ansible_user=centos

[slurm_worker]
compute001 ansible_host=10.0.0.62 ansible_user=centos
compute002 ansible_host=10.0.0.51 ansible_user=centos
```

**Comment by riccardomurri** [comment]
whoops! the `--no-setup` option actually inhibits creation of the inventory file as well...  Please remove it!

To speed things up, you can add this line in the `[setup/*]` section of the cluster:

    # yes, it's `--` + space + `--list-hosts`
    ansible_extra_args = -- --list-hosts

this will make the "setup" part (almost) a no-op.

**Comment by pescobar** [comment]
no luck :(

I get this error when trying to boot the cluster

```
ERROR! the playbook: --list-hosts could not be found
2017-08-17 13:29:12 elasticluster-admin.novalocal gc3.elasticluster[26098] ERROR Command `ansible-playbook /home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/share/playbooks/site.yml --inventory=/home/centos/.elasticluster/storage/slurm_centos7.inventory --become --become-user=root -- --list-hosts` failed with exit code 1.
2017-08-17 13:29:12 elasticluster-admin.novalocal gc3.elasticluster[26098] ERROR Check the output lines above for additional information on this error.
```
this is what I have added to my config

```
[setup/ansible-slurm]
provider=ansible
frontend_groups=slurm_master
compute_groups=slurm_worker
# yes, it's `--` + space + `--list-hosts`
ansible_extra_args = -- --list-hosts
```

**Comment by riccardomurri** [comment]
sorry, my bad -- it's

    ansible_extra_args = --list-hosts

no, prefixed `--` + space.

(I seem not to able to get a command-line right today... :-/)

**Comment by pescobar** [comment]
now I get the same ip for every compute node in the inventory file :)

```
[centos@elasticluster-admin ~]$ cat ~/.elasticluster/storage/slurm_centos7.inventory 

[slurm_master]
frontend001 ansible_host=10.0.0.87 ansible_user=centos

[slurm_worker]
compute001 ansible_host=10.0.0.87 ansible_user=centos
compute002 ansible_host=10.0.0.87 ansible_user=centos
compute003 ansible_host=10.0.0.87 ansible_user=centos
compute004 ansible_host=10.0.0.87 ansible_user=centos
compute005 ansible_host=10.0.0.87 ansible_user=centos
compute006 ansible_host=10.0.0.87 ansible_user=centos
compute007 ansible_host=10.0.0.87 ansible_user=centos
compute008 ansible_host=10.0.0.87 ansible_user=centos
compute009 ansible_host=10.0.0.87 ansible_user=centos
compute010 ansible_host=10.0.0.87 ansible_user=centos
compute011 ansible_host=10.0.0.87 ansible_user=centos
compute012 ansible_host=10.0.0.87 ansible_user=centos
compute013 ansible_host=10.0.0.87 ansible_user=centos
compute014 ansible_host=10.0.0.87 ansible_user=centos
compute015 ansible_host=10.0.0.87 ansible_user=centos
compute016 ansible_host=10.0.0.87 ansible_user=centos
compute017 ansible_host=10.0.0.87 ansible_user=centos
compute018 ansible_host=10.0.0.87 ansible_user=centos
compute019 ansible_host=10.0.0.87 ansible_user=centos
compute020 ansible_host=10.0.0.87 ansible_user=centos
compute021 ansible_host=10.0.0.87 ansible_user=centos
compute022 ansible_host=10.0.0.87 ansible_user=centos
compute023 ansible_host=10.0.0.87 ansible_user=centos
compute024 ansible_host=10.0.0.87 ansible_user=centos
compute025 ansible_host=10.0.0.87 ansible_user=centos
compute026 ansible_host=10.0.0.87 ansible_user=centos
```

**Comment by riccardomurri** [comment]
Please update from the "fix/inventory-with-ipv6" branch and try again.  This time it should really work ;-)

**Comment by pescobar** [comment]
```
2017-08-17 18:12:18 elasticluster-admin.novalocal gc3.elasticluster[26978] ERROR Error: 
Aborting because of errors: .
```

:)

**Comment by riccardomurri** [comment]
AAARRGGHHH!

Any traceback? (You can see it with `elasticluster start -vvvv`) I swear it
works on my laptop :)

Il 17 ago 2017 6:12 PM, "Pablo Escobar Lopez" <notifications@github.com> ha
scritto:

**Comment by pescobar** [comment]
```
(venv-elasticluster-fix-test)[centos@elasticluster-admin venv-elasticluster-fix-test]$ elasticluster -vvvvvvv start slurm_centos7
2017-08-18 09:45:25 elasticluster-admin.novalocal gc3.elasticluster[1780] DEBUG Checking section `cluster/slurm_centos7` ...
2017-08-18 09:45:25 elasticluster-admin.novalocal gc3.elasticluster[1780] DEBUG Checking section `login/centos7` ...
2017-08-18 09:45:25 elasticluster-admin.novalocal gc3.elasticluster[1780] DEBUG Checking section `setup/ansible-slurm` ...
2017-08-18 09:45:25 elasticluster-admin.novalocal gc3.elasticluster[1780] DEBUG Checking section `cloud/switch` ...
2017-08-18 09:45:25 elasticluster-admin.novalocal gc3.elasticluster[1780] ERROR Error: 
Traceback (most recent call last):
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/__main__.py", line 195, in main
    return self.params.func()
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/subcommands.py", line 81, in __call__
    return self.execute()
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/subcommands.py", line 181, in execute
    cluster = creator.load_cluster(cluster_name)
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/conf.py", line 851, in load_cluster
    cluster = repository.get(cluster_name)
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/repository.py", line 365, in get
    return store.get(name)
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/repository.py", line 188, in get
    cluster = self.load(storage)
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/repository.py", line 290, in load
    cluster = Cluster(**data)
  File "/home/centos/venv-elasticluster-fix-test/elasticluster/elasticluster/cluster.py", line 135, in __init__
    assert 'name' in node
AssertionError
Aborting because of errors: .
```

**Comment by riccardomurri** [comment]
Hmmm...  the error does not point to any code that I've touched lately.  It looks like either some file in `~/.elasticluster/storage` is incorrect or the configuration is -- assuming you haven't changed the config from the one that's working, let's go the "corrupt save file" route.

Can you please do:

    elasticluster stop -y slurm_centos7

then try again?

**Comment by pescobar** [comment]
`elasticluster stop -y slurm_centos7` was triggering the same error. Deleting the contents of `$HOME/.elasticluster/storage` workarounds the problem. Now the cluster boots again but the problem is still present (I only had to try once)

This is the debug log https://gist.github.com/pescobar/8652f192785a5f487649c9525ecfdab0
This is the generated inventory: https://gist.github.com/pescobar/a49c761dd0ca434e0bcf923cfd7b5d06

This is the code verstion I am testing:

```
(venv-elasticluster-fix-test)[centos@elasticluster-admin elasticluster]$ git log |head -5
commit 6f258f069bdadd4c3d29eb579980e4a67099c03b
Author: Riccardo Murri <riccardo.murri@gmail.com>
Date:   Thu Aug 17 18:04:27 2017 +0200

    Revise `AnsibleSetupProvider._build_inventory()` thoroughly.
```

**Comment by pescobar** [comment]
btw, I think interesting lines are these ones:

https://gist.github.com/pescobar/8652f192785a5f487649c9525ecfdab0#file-gistfile1-txt-L36
https://gist.github.com/pescobar/8652f192785a5f487649c9525ecfdab0#file-gistfile1-txt-L52
https://gist.github.com/pescobar/8652f192785a5f487649c9525ecfdab0#file-gistfile1-txt-L92

**Comment by riccardomurri** [comment]
> This is the generated inventory: https://gist.github.com/pescobar/a49c761dd0ca434e0bcf923cfd7b5d06

The inventory looks correct to me: you have some hosts with IPv6 addresses, and the address is correctly reported as `ansible_host=...` (instead of `ansible_host=2001 ansible_port=<rest of IPv6 addr>` as it was before).

**Comment by pescobar** [comment]
ups, you are right. I was expecting that the inventory file would only include ipv4 addresses. 

I tried a full boot and ansible could be executed correctly connecting to the ipv6 addresses in the inventory. I have noticed that even if the inventory file includes ipv6 addresses the /etc/hosts created in the compute nodes only includes ipv4 addresses. Is this expected? Why if all nodes have ipv4 address elasticluster is using the ipv6 address for some of them in the inventory?

**Comment by riccardomurri** [comment]
> I have noticed that even if the inventory file includes ipv6 addresses the /etc/hosts created in the compute nodes only includes ipv4 addresses. Is this expected?

It's possible: the inventory and `/etc/hosts` are built in different ways; the inventory is built by ElastiCluster, which tries every IP address returned by OpenStack until one succeeds. The `/etc/hosts` file is built by the Ansible playbooks, which still assume that hosts have an IPv4 address and ignore IPv6.

**Comment by riccardomurri** [comment]
So I think this is good to go.

I will wait for further error reports until Tuesday and merge it into "master" if no new errors happen.

**Comment by pescobar** [comment]
> So I think this is good to go.

I agree. I think it's good to merge

**Comment by riccardomurri** [comment]
Fixed in commit 47c2f7d on "master" branch
